File: Python/SpiralMatrix.py
class SpiralMatrix:
    def spiralOrder(self, matrix):
        if not matrix:
            return []

        n, m = len(matrix), len(matrix[0])
        left, right, top, bottom = 0, m - 1, 0, n - 1
        result = []

        while left <= right and top <= bottom:
            for i in range(left, right + 1):
                result.append(matrix[top][i])
            top += 1

            for i in range(top, bottom + 1):
                result.append(matrix[i][right])
            right -= 1

            if top <= bottom:
                for i in range(right, left - 1, -1):
                    result.append(matrix[bottom][i])
                bottom -= 1

            if left <= right:
                for i in range(bottom, top - 1, -1):
                    result.append(matrix[i][left])
                left += 1

        return result

# Example usage
spiral = SpiralMatrix()
matrix = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]
result = spiral.spiralOrder(matrix)
print(result)

File: JS/SpiralMatrix.js
class SpiralMatrix {
    spiralOrder(matrix) {
        const n = matrix.length;
        const m = matrix[0].length;
        let left = 0, right = m - 1;
        let top = 0, bottom = n - 1;
        const result = [];

        while (top <= bottom && left <= right) {
            for (let i = left; i <= right; i++) {
                result.push(matrix[top][i]);
            }
            top++;

            for (let i = top; i <= bottom; i++) {
                result.push(matrix[i][right]);
            }
            right--;

            if (top <= bottom) {
                for (let i = right; i >= left; i--) {
                    result.push(matrix[bottom][i]);
                }
                bottom--;
            }

            if (left <= right) {
                for (let i = bottom; i >= top; i--) {
                    result.push(matrix[i][left]);
                }
                left++;
            }
        }

        return result;
    }
}

// Example usage
const spiral = new SpiralMatrix();
const matrix = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
];
const result = spiral.spiralOrder(matrix);
console.log(result);
